test/e2e/case23_compliance_api_recording_test.go

```diff
@@ -0,0 +1,266 @@
+			r.Header.Set("Content-Type", "application/json")
+			if r.Method != http.MethodPost {
+				w.WriteHeader(http.StatusMethodNotAllowed)
+				_, _ = w.Write([]byte(`{"message": "Invalid method"}`))
+
+				return
+			}
+
+			body, err := io.ReadAll(r.Body)
+			if err != nil {
+				log.Error(err, "error reading request body")
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(`{"message": "Could not read request body"}`))
+
+				return
+			}
+
+			reqEvent := &complianceeventsapi.ComplianceEvent{}
+
+			if err := json.Unmarshal(body, reqEvent); err != nil {
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(`{"message": "Invalid JSON"}`))
+
+				return
+			}
+
+			if err := reqEvent.Cluster.Validate(); err != nil {
+				log.Error(err, "Invalid cluster")
+
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(`{"message": "Invalid cluster"}`))
+
+				return
+			}
+
+			if err := reqEvent.Event.Validate(); err != nil {
+				log.Error(err, "Invalid event")
+
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(`{"message": "Invalid event"}`))
+
+				return
+			}
+
+			if reqEvent.Policy.KeyID == 0 {
+				w.WriteHeader(http.StatusBadRequest)
+				_, _ = w.Write([]byte(`{"message": "Invalid policy ID"}`))
+
+				return
+			}
+
+			w.WriteHeader(http.StatusCreated)
+
+			lock.Lock()
+			requests = append(requests, reqEvent)
+			lock.Unlock()
+		})
+
+		go func() {
+			defer GinkgoRecover()
+
+			err := server.ListenAndServe()
+			Expect(err).To(MatchError(http.ErrServerClosed))
+		}()
+	})
+
+	AfterAll(func(ctx context.Context) {
+		err := server.Shutdown(ctx)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Deleting a policy on hub cluster in ns:" + clusterNamespaceOnHub)
+		_, err = kubectlHub("delete", "-f", yamlPath, "-n", clusterNamespaceOnHub, "--ignore-not-found")
+		Expect(err).ToNot(HaveOccurred())
+		opt := metav1.ListOptions{}
+		utils.ListWithTimeout(clientHubDynamic, gvrPolicy, opt, 0, true, defaultTimeoutSeconds)
+		utils.ListWithTimeout(clientManagedDynamic, gvrPolicy, opt, 0, true, defaultTimeoutSeconds)
+
+		By("clean up all events")
+		_, err = kubectlManaged("delete", "events", "-n", clusterNamespace, "--all")
+		Expect(err).ShouldNot(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		lock.Lock()
+		defer lock.Unlock()
+
+		requests = []*complianceeventsapi.ComplianceEvent{}
+	})
+
+	It("Forwards compliance events from a controller to the compliance API", func(ctx context.Context) {
+		hubApplyPolicy("case23", yamlPath)
+
+		By("Generates a Pending message")
+		generateEvent(ctx, "case23", "case23", "Halt, who goes there?", "Pending")
+		By("Generates a NonCompliant message")
+		generateEvent(ctx, "case23", "case23", "You shall not pass", "NonCompliant")
+		By("Generates a Compliant message")
+		generateEvent(ctx, "case23", "case23", "You may pass", "Compliant")
+
+		By("Waiting for the compliance API requests to come in")
+		Eventually(
+			func(g Gomega) {
+				lock.RLock()
```

Comment:
Why this is needed?

Reply:
Because I don't want to access `requests` if the HTTP server is writing to the `requests` slice.